pytensor/link/pytorch/dispatch/scalar.py

```diff
@@ -38,3 +37,32 @@ def pytorch_func(*args):
             )
 
     return pytorch_func
+
+
+@pytorch_funcify.register(ScalarLoop)
+def pytorch_funicify_ScalarLoop(op, node, **kwargs):
+    update = pytorch_funcify(op.fgraph)
+    if op.is_while:
+
+        def scalar_loop(steps, *start_and_constants):
+            *carry, constants = start_and_constants
+            constants = constants.unsqueeze(0)
+            done = True
+            for _ in range(steps):
+                *carry, done = update(*carry, *constants)
+                constants = start_and_constants[len(carry) :]
+                if done:
+                    break
+            return torch.stack((*carry, done))
```

Comment:
I'm not super happy about this since we're mixing dtypes (boolean to int isn't the worst thing in the world) but that's due to the torch.cpu thing that happens in the linker.

Reply:
We shouldn't need to stack, is the problem due to `compare_pytorch_and_py` issue? I have changed the code in other PRs still not merged to look like this: 
![image](https://github.com/user-attachments/assets/57f079ee-0846-41cf-bb7b-98028c08471a)

Reply:
I thought it was from the linker, but it could be the test! Let me try it again with the change you made.

Reply:
Yeah, @ricardoV94 the issue is the linker expects all the results to be tensors https://github.com/pymc-devs/pytensor/blob/main/pytensor/link/pytorch/linker.py#L15-L16

`update` returns a tuple, we unpack it when we call update, but that unpack doesn't happen on return. Stack is one way to make all the values a tensor; there might be another. Any suggestions?

Big ole stack tracedump

```
=========================================================== FAILURES ===========================================================
_______________________________________________________ test_ScalarLoop ________________________________________________________

    def streamline_default_f():
        for x in no_recycling:
            x[0] = None
        try:
            for thunk, node, old_storage in zip(
                thunks, order, post_thunk_old_storage
            ):
>               thunk()

pytensor/link/utils.py:196: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pytensor/link/basic.py:669: in thunk
    o_storage[0] = self.output_filter(o_var, o_val)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pytensor.link.pytorch.linker.PytorchLinker object at 0x11f7fe450>, var = ScalarLoop.0
out = (tensor(5., dtype=torch.float64),)

    def output_filter(self, var: Variable, out: Any) -> Any:
>       return out.cpu()
E       AttributeError: 'tuple' object has no attribute 'cpu'

pytensor/link/pytorch/linker.py:16: AttributeError

During handling of the above exception, another exception occurred:

    def test_ScalarLoop():
        n_steps = int64("n_steps")
        x0 = float64("x0")
        const = float64("const")
        x = x0 + const
    
        op = ScalarLoop(init=[x0], constant=[const], update=[x])
        x = op(n_steps, x0, const)
    
        fn = function([n_steps, x0, const], x, mode=pytorch_mode)
>       np.testing.assert_allclose(fn(5, 0, 1), 5)

tests/link/pytorch/test_basic.py:318: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pytensor/compile/function/types.py:959: in __call__
    self.vm()
pytensor/link/utils.py:200: in streamline_default_f
    raise_with_op(fgraph, node, thunk)
pytensor/link/utils.py:524: in raise_with_op
    raise exc_value.with_traceback(exc_trace)
pytensor/link/utils.py:196: in streamline_default_f
    thunk()
pytensor/link/basic.py:669: in thunk
    o_storage[0] = self.output_filter(o_var, o_val)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pytensor.link.pytorch.linker.PytorchLinker object at 0x11f7fe450>, var = ScalarLoop.0
out = (tensor(5., dtype=torch.float64),)

    def output_filter(self, var: Variable, out: Any) -> Any:
>       return out.cpu()
E       AttributeError: 'tuple' object has no attribute 'cpu'
E       Apply node that caused the error: ScalarLoop(n_steps, x0, const)
E       Toposort index: 0
E       Inputs types: [ScalarType(int64), ScalarType(float64), ScalarType(float64)]
E       Inputs shapes: [(), (), ()]
E       Inputs strides: [(), (), ()]
E       Inputs values: [5, 0.0, 1.0]
E       Outputs clients: [[output[0](ScalarLoop.0)]]
```

Reply:
The problem is this line in the test: `x = op(n_steps, x0, const)`. It should be unpacked into multiple variables

Reply:
resolved this with updating the linker